### ISSUE TITLE
DEV API returning wrong results when querying articles with tags

### ISSUE THREAD

**Issue by remotesynth** [forem/forem#12702]
I am trying to use the DEV API to query [published articles](https://docs.dev.to/api/#operation/getArticles). Here are some odd inconsistencies that are either not working as they should or not properly documented:

* Combining a query with `username` and `tag` parameters will ignore the username and send back articles for all.
* The `tags` attribute doesn't seem to have any effect at all as the articles returned do not have the tags that are specified

There doesn't appear a way, given the above, to query a user's articles that have a specific tag.

**To Reproduce**

`https://dev.to/api/articles?username=remotesynth&tag=discuss`

The above will return articles for all users.

`https://dev.to/api/articles?username=remotesynth&tags=discuss, flash`

(Note: I chose those tags because they should return few results)

The above will return all articles for remotesynth.

`https://dev.to/api/articles?tags=discuss, flash`

The above will return all articles for all users with any tags.

Note that I have tried all variations of the `tags` attribute - with comma and space (as specified in the docs), with no space, or even just a single tag. None of the variations seemed to work.

**Expected behavior**

`https://dev.to/api/articles?username=remotesynth&tag=discuss`

Should return articles for remotesynth with the tag discuss.

`https://dev.to/api/articles?username=remotesynth&tags=discuss, flash`

Should return only articles for remotesynth that are tagged either discuss or flash.

`https://dev.to/api/articles?tags=discuss, flash`

Should return articles for any user that are tagged either discuss or flash.

**Comment by rhymes** [comment]
Hi @remotesynth, thanks for the report!

> Combining a query with username and tag parameters will ignore the username and send back articles for all.

This will currently favor the tag over the username param. So `https://dev.to/api/articles?username=remotesynth&tag=discuss` is essentially the same as `https://dev.to/api/articles?tag=discuss`

The [documentation](https://docs.dev.to/api/#operation/getArticles) mentions that `tag` can only be used together with `top` and that `username` can only be used with the `state` param.

There's currently no way to get "public articles of a user tagged with something".

> The tags attribute doesn't seem to have any effect at all as the articles returned do not have the tags that are specified

The `tags` attribute can only be used by itself, but it's true, there's a bug as I can't use `tags` or `tags_exclude` either. It's probably related to caching.

**TLDR;**:

- the first is a known behavior
- the second is definitely a bug

**Comment by vishaldeepak** [comment]
looks interesting, any directions on how to make an attempt on figuring out problems related to caching on local server?

**Comment by mikeydorje** [comment]
Closing as stale. No activity since August 2021.

Based on code review and testing, the `tags` and `tags_exclude` parameters now function correctly - specs pass and the ArticleApiIndexService properly handles these parameters. While rhymes confirmed in Feb 2021 that these parameters were "definitely a bug" (possibly caching-related), the functionality appears to have been resolved at some point over the past 5 years.

The first issue (username+tag combination) remains a known limitation by design - parameters are mutually exclusive.

If the `tags`/`tags_exclude` parameters are still not working correctly on current DEV or your instance, please reopen with updated details and reproduction steps.
